enhancements/machine-config/admin-defined-node-disruption-policy.md

```diff
@@ -0,0 +1,311 @@
+
+NodeDisruptionPolicies are only processed in-cluster and cannot be used to skip the bootstrap pivot during cluster installation.
+
+#### Variation and form factor considerations [optional]
+
+This will work by default on standalone OCP and any other OCP variants that run the MCO in its entirety. (e.g. works for RHEL)
+
+HyperShift deployments do not run the MCO pods directly and will not have MachineConfiguration objects.
+
+### API Extensions
+
+- Adds fields to the MachineConfiguration CRD
+
+### Implementation Details/Notes/Constraints [optional]
+
+This aims more to be a user-facing API to essentially map MachineConfig fields to what action should be taken on them, which aligns with existing MCO functionality to perform a hard-coded list of these policies. However, as we evolve the MCO to fit in with an image based OS update vision, we would need this to work with:
+ - Existing MachineConfig based entries
+ - User-provided base-image upgrade cases (live-apply)
+ - A more unified update interface with bootc
+
+Which we can definitely translate under-the-hood to adapt to the system, but file/unit specific entries like this may not work as effectively when it comes to image-to-image updates, especially at the daemon level.
+
+As there is a need for this in 4.16, we will proceed with the format that works best with the existing MCO architecture, while leaving it open to change for the future.
+
+#### Hypershift [optional]
+
+As noted above, Hypershift cannot directly tap into this. In a bit more detail, Hypershift has 2 modes of operation for nodepools:
+Replace upgrades - since all machines are deleted and recreated in this model, there is no method of live-application today
+Inplace upgrades - these actually use a special form of the daemon, and we can leverage that to update and perform necessary policies. However we probably do not want hosted cluster admins to have direct control over this, much like MachineConfigs, so if we do want to add this we will need a new API in Hypershift to handle this. As such, we are not considering this at this time
+
+### Risks and Mitigations
+
+As noted in the non-goal sections, any NodeDisruptionPolicy is not verified for application correctness in the MCO, minus formatting. This means that if you accidentally define a change that needed a reboot to take effect, the cluster would “finish the update” without being able to use the new configuration.
+
+The user thereby needs to understand that no additional verification is taking place, and that they accept any responsibility for NodeDisruptionPolicies. If any are applied to the cluster, a metric should be raised for debuggability.
+
+Given the popularity of using pause on MachineConfigPools, it is possible some users will exploit this to skip upgrade disruption. That will likely not break the cluster but will put the configuration in an awkward spot, as we can no longer guarantee that the configuration a node claims to be on is actually applied.
```

Comment:
I didn't think the disruption policy would apply to cluster version upgrades?

Reply:
It doesn't. Removed